### PR TITLE
[WIP] Test Controller Service Plans using mocks and adding gomock to the project.

### PR DESCRIPTION
**Pull request by n3wscott**
This moves controller_serviceplan.go from 42.4% to 93.9% test coverage.

### CODE REVIEW

**Comment by MHBauer** [comment]
Do we have any way of tracking this coverage over any period of time? I'd be more interested in seeing that than increase the coverage of such a small file by adding 4k lines of library.

**Comment by n3wscott** [comment]
My coverage changes are from running the tool and looking at the report before and after.

I also had to fix the tool, there was a path issue. That is included in this PR.

**Comment by vaikas-google** [comment]
@MHBauer you do understand that these mocks can then be used in the subsequent test modifications? This PR is showing how we can make more targeted unit tests by utilizing the mock libraries.
I think having coverage stats would be good though.

**Comment by MHBauer** [comment]
I understand it can be used in other places.

Is there a comparative analysis we can do? Write the same test with both fakes and mocks and compare?

**Comment by n3wscott** [comment]
Fakes do not exist for the listers so someone would have to write that. Which is what gomock is doing for us in that 4k lines of code, except it can do it for anything... If used correctly this will be powerful.

**Comment by staebler** [comment]
I'm inclined to agree with @MHBauer on comparative analysis. I don't see anything in these changes that we can't already do with our existing fakes. The listers can be set up to do what is being done in `TestReconcileClusterServicePlanKeyThrowsError` and `TestReconcileClusterServicePlanKeyNotFound` by simple adding the necessary service plans to the informers returned from `getTestController`. The increased test coverage is not a consequence of adding mocking: it is a consequence of adding more tests.

**Comment by n3wscott** [comment]
> by simple adding the necessary service plans to the informers

This is what I am trying to avoid. To be able to get to the error state you need to understand the workings of the dependent components. I don't want to have to know how to set-up a lister to return the wanted state.

Again, these are simple versions of usage.  Perhaps I need to add a more complicated test that uses the mocks to show its value.

**Comment by pmorie** [comment]
I'm not sure about this one and want a chance to think about it.

**Comment by n3wscott** [comment]
TODO: Add an example with the new refactored instance delete.

**Comment by arschles** [comment]
I believe I've said this before, but not sure so I'll repeat here.

I don't see any risk in adding this change right now to see how the experience of writing tests improves (or doesn't) long-term.

We have a rough idea of how coverage improves now, but I also think tracking coverage stats would be good over time. Again, something we can do in a follow-up.

File: pkg/client/listers_generated/servicecatalog/v1beta1/mocks_generate.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package v1beta1
```
Comment (staebler) [inline pkg/client/listers_generated/servicecatalog/v1beta1/mocks_generate.go]: Is this file manually created? I am unclear how the mock lister is generated. If this file is manually created, then it is in conflict with the warning [here](https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/client/WARNING.md).

**Pull request closed without merging**